xarray/tutorial.py

```diff
@@ -6,23 +6,129 @@
+    md5.update(content)
+    md5sum = md5.hexdigest()
+
+    return md5sum == checksum
+
+
+# based on https://stackoverflow.com/a/29491523
+@contextmanager
+def open_atomic(path, mode=None):
+    folder = path.parent
+    prefix = f".{path.name}"
+
+    try:
+        f = tempfile.NamedTemporaryFile(dir=folder, prefix=prefix, delete=False)
+        temporary_path = pathlib.Path(f.name)
+
+        yield f
+
+        temporary_path.rename(path)
+    except Exception:
+        # switch to path.unlink(missing_ok=True) for Python 3.8+
+        with suppress(FileNotFoundError):
+            temporary_path.unlink()
+        raise
+
+
+def download_to(url, path):
+    # based on https://stackoverflow.com/a/39217788
+    with open_atomic(path, mode="wb") as f:
+        with requests.get(url, stream=True) as r:
+            if r.status_code != 200:
+                raise OSError(f"download failed: {r.reason}")
+
+            r.raw.decode_content = True
+            shutil.copyfileobj(r.raw, f)
+
+
+def open_rasterio(
+    name,
+    cache=True,
+    cache_dir=_default_cache_dir,
+    github_url="https://github.com/mapbox/rasterio",
+    branch="master",
+    **kws,
+):
+    """
+    Open a rasterio dataset from rasterio's online repository (requires internet).
+
+    If a local copy is found then always use that to avoid network traffic.
+
+    Parameters
+    ----------
+    name : str
+        Name of the file containing the dataset. If no suffix is given, assumed
+        to be TIF ('.tif' is appended)
+        e.g. 'RGB.byte'
+    cache_dir : path-like, optional
+        The directory in which to search for and write cached data.
+    cache : boolean, optional
+        If True, then cache data locally for use on subsequent calls
+    github_url : string
+        Github repository where the data is stored
+    branch : string
+        The git branch to download from
+    kws : dict, optional
+        Passed to xarray.open_dataset
+
+    See Also
+    --------
+    xarray.open_rasterio
+    xarray.tutorial.open_dataset
+
+    """
+    if isinstance(cache_dir, str):
+        cache_dir = pathlib.Path(cache_dir)
+
+    if not cache_dir.is_dir():
+        cache_dir.mkdir()
+
+    default_extension = ".tif"
+
+    if cache:
+        path = cache_dir / name
+        # need to always do that, otherwise we might close the
+        # path using the context manager
+        cache_dir = pathlib.Path(cache_dir)
+    else:
+        cache_dir = tempfile.TemporaryDirectory()
+        path = pathlib.Path(cache_dir.name) / name
+
+    if not path.suffix:
+        path = path.with_suffix(default_extension)
+    elif path.suffix == ".byte":
+        path = path.with_name(name + default_extension)
+
+    if cache and path.is_file():
+        return _open_rasterio(path, **kws)
+
+    url = f"{github_url}/raw/{branch}/tests/data/{path.name}"
+    # if cache_dir points to a temporary directory, make sure it is
+    # deleted afterwards
+    with cache_dir:
```

Comment:
I feel like I'm missing something -- what does using a pathlib.Path object as a context manager do?

I don't think it automatically deletes the files after using it?

Reply:
no, of course not. `cache_dir` can be either a `pathlib.Path` instance or a `NamedTemporaryDirectory` instance. Using a `Path` object as a context manager will "close" it so you can't use it for I/O anymore (`unlink`, `read_text`, etc. won't work anymore) while a `NamedTemporaryDirectory` instance will delete the temporary directory it created when `__exit__` is called.

Not sure if there's a way to make that easier to read.